src/mono/mono/utils/mono-wasm-pagemgr.c

```diff
@@ -0,0 +1,409 @@
+	g_assert (first_page < MWPM_MAX_PAGES);
+
+	uint32_t last_page = first_page + (page_count - 1);
+	g_assert (last_page >= first_page);
+
+	g_assert (last_page < MWPM_MAX_PAGES);
+
+	// POSIX specifies that munmap () on an address range that isn't mapped has no,
+	//  effect, so we need to make sure that it's harmless to try and unmap pages we
+	//  don't control. We can't use memset since it might trample UNKNOWN pages.
+	for (uint32_t i = first_page, skip_value = page_count - 1; i <= last_page; i++) {
+		mwpm_page_state page_state = page_table[i];
+
+		// TODO: Remove the duplication in here
+		switch (action) {
+			case MWPM_MARK_DEAD_PAGES:
+				g_assert (!is_page_owned (page_state));
+				page_table[i] = encode_page_state (MWPM_EXTERNAL, skip_value--);
+				break;
+			case MWPM_MARK_NEW_PAGES:
+				g_assert (!is_page_owned (page_state));
+				page_table[i] = encode_page_state (MWPM_FREE_ZEROED, skip_value--);
+				break;
+			case MWPM_FREE_TO_ALLOCATED:
+				g_assert (is_page_free (page_state));
+				page_table[i] = encode_page_state (MWPM_ALLOCATED, skip_value--);
+				break;
+			case MWPM_FREE_TO_ALLOCATED_ZEROED:
+				g_assert (is_page_free (page_state));
+				page_table[i] = encode_page_state (MWPM_ALLOCATED, skip_value--);
+				if (!(page_state & MWPM_META_BIT))
+					// TODO: Don't recalculate the address from scratch each time
+					memset (address_from_page_index (i), 0, MWPM_PAGE_SIZE);
+				break;
+			case MWPM_ALLOCATED_TO_FREE:
+				// FIXME: Can we generate correct skip_value here? This is used
+				//  by munmap, which is valid to call even on pages that are not mapped
+				if (is_page_in_use (page_state))
+					page_table[i] = encode_page_state (MWPM_FREE_DIRTY, 0);
+				break;
+			default:
+				g_assert_not_reached ();
+				break;
+		}
+	}
+}
+
+static void
+print_stats () {
+#if defined(MWPM_LOGGING) || defined(MWPM_STATS)
+	uint32_t in_use = 0, free = 0, unallocated = 0,
+		max_run = 0, current_run = 0;
+
+	for (uint32_t i = first_controlled_page_index; i <= last_controlled_page_index; i++) {
+		switch (page_table[i] & MWPM_STATE_MASK) {
+			case MWPM_ALLOCATED:
+				in_use++;
+				current_run = 0;
+				break;
+
+			case MWPM_FREE_DIRTY:
+			case MWPM_FREE_ZEROED:
+				free++;
+				current_run++;
+				if (current_run > max_run)
+					max_run = current_run;
+				break;
+
+			default:
+				unallocated++;
+				current_run = 0;
+				break;
+		}
+	}
+
+	uint32_t total = in_use + free; // + unallocated;
+	g_print (
+		"sbrk(0)==%u. %u pages in use (%f%%), %u pages free, %u pages unknown. largest possible allocation: %u pages\n",
+		(uint32_t)sbrk(0), in_use, in_use * 100.0 / total, free, unallocated, max_run
+	);
+#endif
+}
+
+static void *
+acquire_new_pages_initialized (uint32_t page_count) {
+	if (page_count < 1)
+		return NULL;
+	uint64_t bytes = (page_count + 1) * MWPM_PAGE_SIZE;
+	uint32_t recovered_bytes = 0;
+	if (bytes >= UINT32_MAX)
+		return NULL;
+	// We know that on WASM, sbrk grows the heap by a set number of pages,
```

Comment:
Should we try to see if we can get `page_count` pages into `prev_waste_end - prev_waste_start` before we call `sbrk`?

Reply:
We don't know until we call sbrk whether we can recover the waste

Reply:
I thought everything between `prev_waste_start` and `prev_waste_end` is memory that we got from a previous call to `sbrk` that we now own.  If there was a foreign caller of `sbrk` they would start allocating from at least `prev_waste_end` (the last `allocation_end` that we recorded).  Did I misunderstand?

Reply:
Oh, I see. Well, we know by definition the waste is smaller than a page, so we can't use it.

Reply:
I don't understand why we attempt to return reclaimed waste to the caller then.  aren't they expecting something aligned? Or do they have looser alignment requirements (is this the 16kb vs 64kb page thing?)

Also if we do successfully return to part of the reclaimed waste to a caller are we in trouble when it comes time to unmap?  Don't we now have two allocations that both think they own that address range?  The first allocation request (that caused the waste to be created) only owns up to `allocation_end`.  So when they call unmap, do we somehow know that only part of the page should become unmapped?

Reply:
The way the waste handling works is we only use it if we can turn waste into a complete page. This is only possible if we make two sbrk() calls back-to-back. i.e.

sbrk call 1 produces 8KB of waste at the front and 56KB of waste at the end. Neither of these can be used to create a page, they're too small.
sbrk call 2 produces another 8KB of waste at the front and 56KB of waste at the end. The 8KB of waste at the front is adjacent to the 56KB from sbrk call 1, so we can weld those two pieces of waste together into a usable page.

Reply:
Sorry, maybe I'm being dense here - after sbrk call 2 if we return an address range starting from the 56kb of waste and then immediately pass that address to `mwpm_free_range` won't `first_page_from_address` round down and we'll end up unmapping the pages that were acquired by sbrk call 1?

---

Also, we keep track of whether memory is free/used/zero/dirty in aligned 64kb chunks, but it seems like `acquire_new_pages_initialized` can return un-aligned memory by taking advantage of the waste or if Emscripten `sbrk` gives us back an unaligned result?  But we then try to return 64k-aligned pages from `mwpm_alloc_range`, right?

So don't you need to round up the allocation request to `sbrk` to fix up the alignment?

something like:
```c
   void* current_end = sbrk(0);
   uint64_t misaligned_frontier_size = (MWPM_PAGE_SIZE - (((uint32_t)current_end)%MWPM_PAGE_SIZE);
   void * allocation = sbrk ((uint32_t)(bytes + misaligned_frontier_size));
   g_assert (allocation == current_end);
   allocation += misaligned_frontier_size;
  // now we know allocation is 64kb aligned and it extends for `bytes` bytes
```

Reply:
We always return aligned addresses and don't allocate partial pages, unless I screwed something up. The sbrk(0) idea is a race condition, so I opted not to do it (even though in the ideal case it would reduce waste).

We do round up the allocation request to sbrk, by adding an extra page to it. 
```c
uint64_t bytes = (page_count + 1) * MWPM_PAGE_SIZE;
```

This extra page worth of space will get split across both sides of the allocation, then the tail of it will be recovered if we're the next person to call sbrk.